### PR TITLE
Fetch status from either headers or trailers

### LINKED ISSUE #1: Support "trailers-only" responses
This is possibly related to the hang-forever bug reported by @mbacarella in dialohq/ocaml-grpc-fork#2.

Servers can respond with an empty body and headers containing the gRPC status that would normally be packaged as trailing headers at the end of the body. E.g. if you call `etcd` with a non-existing method, the response is just HTTP headers:

```
((":status" "200")("content-type" "application/grpc")("grpc-status" "12")
("grpc-message" "unknown method This_method_is_missing"))
```
And the Lwt client will hang.  (Status code 12 is UNIMPLEMENTED)

Confusingly, this is called "Trailers-only" in the [gRPC spec](https://github.com/grpc/grpc/blob/master/doc/PROTOCOL-HTTP2.md#responses).
 
How to fix this:
- Check for gRPC status in the headers first
- If they're missing, wait for trailers in the body
- Only when both are missing, apply @mbacarella's workaround (in this case, we probably dialed a non-gRPC server)

### PR DESCRIPTION
**Pull request by quernd**
Contribution by @doctor-pi (see https://github.com/dialohq/ocaml-grpc-fork/pull/5) that should be migrated to the new repo.

Closes https://github.com/dialohq/ocaml-grpc/issues/1

### CODE REVIEW

File: lib/grpc-lwt/client.ml [hunk 1]
```diff
@@ -48,9 +37,24 @@ let call ~service ~rpc ?(scheme = "https") ~handler ~do_request
...
+    Lwt.wakeup_later read_body_notify body;
+    Lwt.async (fun () ->
+        if response.status <> `OK then (
+          Lwt.wakeup_later out_notify
+            (Error (Grpc.Status.v Grpc.Status.Unknown));
+          Lwt.return_unit)
+        else
+          let+ handler_res in
+          Lwt.wakeup_later out_notify (Ok handler_res));
+    Lwt.async(fun () -> Lwt.return (trailers_handler response.headers))
```
Comment (quernd) [inline lib/grpc-lwt/client.ml]: `Lwt.async` seems unnecessary here:
```suggestion
    trailers_handler response.headers
```

File: lib/grpc-async/client.ml [hunk 2]
```diff
@@ -49,10 +37,26 @@ let call ~service ~rpc ?(scheme = "https") ~handler ~do_request
...
+    don't_wait_for
+      (match response.status with
+      | `OK ->
+          let%bind handler_res = Ivar.read handler_res_ivar in
+          Ivar.fill out_ivar (Ok handler_res);
+          return ()
+      | _ ->
+          Ivar.fill out_ivar (Error (Grpc.Status.v Grpc.Status.Unknown));
+          return ());
+    don't_wait_for (return (trailers_handler response.headers))
```
Comment (quernd) [inline lib/grpc-async/client.ml]: `don't_wait_for` seems unnecessary here:
```suggestion
    trailers_handler response.headers
```

File: lib/grpc-lwt/client.ml (hunk 1 above)
Comment (dimitris-m) [reply lib/grpc-lwt/client.ml]: I think you're right, let me double-check just in case.

**Comment by dimitris-m** [comment]
Agree with you on both changes. 
I sent you a patch on discord, for convenience.

**Comment by quernd** [comment]
Thanks, patch applied! @wokalski let's merge this?

**Review by wokalski — APPROVED** [review]

**Comment by wokalski** [comment]
No need to wait for me next time. 

@doctor-pi You're a collaborator so you could've pushed to the branch directly. Just FYI :)

**Comment by dimitris-m** [comment]
> No need to wait for me next time.
> 
> @doctor-pi You're a collaborator so you could've pushed to the branch directly. Just FYI :)

And miss my chance to send an old school email patch? :) 

But you are right. Thanks!

**Pull request merged**